[Http] General design, structure and convention guidelines (#226558)
elastic/kibana: dev_docs/contributing/kibana_http_api_guidelines.mdx

**Do contain version numbers**

`/api/my_api` ✅

`/api/v1/my_api` ❌

See the section on [versioning](#versioning) for more details.

**For public APIs, prefix with `/api`**

`/api/my_api` ✅

`/my_api` ❌

**Pluralize collection names**

`PUT /api/resources/{id}` ✅

`PUT /api/resource/{id}` ❌

### Methods

**GET does not accept a body**

`GET /api/resources/{id}` ✅

`GET /api/resources/{id} { body: { params } }` ❌

**POST is for creating a resource**

Sometimes `POST` can be used for an action against a resource or for actions like `GET` but with a body.

**PUT or PATCH are for updating a resource**

`PATCH` is typically used for partial updates and should not result in something being created, while `PUT` might create a resource.

**DELETE is for deleting a resource**

Delete can return a simple response instead of the full resource:

```
DELETE /api/resources/{id}
# -> 204 No content
```

**Stick to simple methods**

`GET`, `POST`, `PUT`, `PATCH`, `DELETE` should cover almost all cases. If you are considering `HEAD` or some other method make sure you have a good justification for your HTTP API.

### Path parameters

**Should use snake case**

`/api/my_api/{my_id}` ✅

> `/api/my_api/{myId}` ❌

Author: That's a good point. Yes it was mainly for the documentation. I'm going to shift sides to "use snake case in bodies", so requiring them in paths should be OK.

Accepting very long strings in query parameters can cause issues for HTTP servers that limit the byte length of certain parts of requests.

Use a body parameter for long strings instead and consider using validation to limit a query param's length.

### Bodies

**Should contain camelCase keys**

While this diverges from the path, in JavaScript the convention to use `camelCase`
in our request and response bodies provides a better developer experience while
maintaining a solid API user experience.

**Use JSON**

Both requests and responses should be `application/json`, unless there is a good justification to use a different media type like when you're serving a file to a client.

**Resource shapes should be consistent**

`GET`, `PUT`, `POST` should return the same shape of data for the same resource.

`GET /api/my_resources/{id}` <br />
=> <br />
`{ "id": "1", "name": "My resource" }`<br /><br />

`POST /api/my_resources/{id} { "id": "1", "name": "My resource" }`<br />
=> <br />
`{ "id": "1", "name": "My resource" }`<br /><br />

`GET /api/my_resources` <br />
=> <br />
`{ items: [{ "id": "1", "name": "My resource" }], page: 1, size: 10, total: 100 }`

> ### Defaults

Author: I this could be relevant for the TF guidelines yeah!

**Should not be changed lightly**

Changing the value of a default may, in some cases, have a devastating result similar to a breaking change.

### Validation

**Runtime validation should be as narrow as feasible**

`schema.object({ id: schema.string({ minLength: 32, maxLength: 32 }) })` ✅

`schema.object({ id: schema.string() })` ❌

**It is easier to relax requirements than tighten them**

If you are in doubt, rather go with stricter validation. Make a requirement more lax if needed is never a breaking change!

### Headers

**Should not be used to specify behaviour**

Outside of exceptional cases, you should always use the parameters, query parameters or body of the request to specify behaviour.

### Response codes

**Should always be used as close as possible to their semantic meaning**

See the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status).

### Pagination and sorting

**Page-based request**

`GET /api/resources?page=1&size=10`

**Page-based response**

```json
{
  "items": [...],
  "total": 100,
  "page": 1,
  "size": 10
}
```

**Cursor based request**

`GET /api/resources?pit_id=abc&search_after="valueA.valueB"&size=10`

> Where `valueA` and `valueB` are joined sort values identifying the last hit. See [ES docs](https://www.elastic.co/docs/reference/elasticsearch/rest-apis/paginate-search-results#search-after). Note: you may not need to support cursor based pagination if your resource will not be in excess of 1000 instances.

Author: Good point! I'll update.

**Cursor based response**

```json
{
  "items": [...],
  "total": 100,
  "pit": "abc",
  "search_after": "valueA1.valueB1",
  "size": 10
}
```

**Pagination requires sorting**

Carefully decide a default for sorting resources returned by any API that support pagination. It is possible to accept a custom
set of additional values for sorting, but this may not be needed. Adding this later is a non breaking change!

Specify sorting in the query parameters like:

```

> GET /api/my_resources?sort=field_name,-other_name

Author: Good point!

I'm also happy to change this asc desc thing if folks disagree with the formatting. I've seen `asc(fieldName)` being recommended by obs devex

  To further reinforce these different purposes Kibana HTTP APIs serve, we categorize them as either **internal** or **public**. See the [Internal vs Public APIs](#internal-vs-public-apis) section for more details.
</DocCallOut>

### Commitment

Public HTTP APIs are 18+ month commitments to users. We must maintain them and ensure they work as expected.

First release new public HTTP APIs as `Technical Preview` or `Beta` to derisk any aspects of the API design you are uncertain about.

Getting your HTTP APIs into a stable state is a process based on feedback from actual users.

#### Breaking changes to HTTP APIs include

 - Rename or delete an API
 - Rename or delete a path, query or body parameter
 - Modifying the type of a property (including expanding to a union type!)
 - Set an existing property as required
 - Add or delete a security requirement
 - In some cases, changing a default

<DocCallOut title="Do not break clients!">
  A public HTTP API should never cause a client to break. Not without a long (18 month) deprecation period and a ready alternative.

  Even `Technical Preview` and `Beta` HTTP APIs should consider graceful paths for changes when possible.

>   Linus Torvalds famously said of Kernel development: "WE DO NOT BREAK USERSPACE!" (along with some other expletives). We should adopt this kind of rigor and empathy when working with our _public_ HTTP APIs and always prioritize finding ways to avoid breaking our APIs.

Reviewer: It would be nice to add a ref here

</DocCallOut>

## Internal vs Public APIs

Public HTTP APIs are those that are exposed to users and are used by plugins and external tools. They are the focus of this document.

Internal HTTP APIs are only used by code Elastic owns, like the UI. By default, when you register a new HTTP API it will be classified as `internal`. For example, this API is internal:

```typescript
router.get({ path: '/api/foos' ... }, async () => {...})
```

You can make an HTTP API public by changing the `access` setting to `public`:

```typescript
router.get({ path: '/api/foos', options: { access: 'public' } }, async () => {...})
```

See the principle regarding [commitment](#commitment) before making APIs public!

## Structure and conventions

The following are a quick reference for how to structure your HTTP APIs.

> ### RESTful-like APIs

Author: Please weigh in here, is this point helpful? Shall I remove it?

Reviewer: It's useful in the sense that it gives contributors a base spec to fall back to if they're completely lost about how to start.

```
DELETE /api/resources/{id}
# -> 204 No content
```

**Stick to simple methods**

`GET`, `POST`, `PUT`, `PATCH`, `DELETE` should cover almost all cases. If you are considering `HEAD` or some other method make sure you have a good justification for your HTTP API.

### Path parameters

**Should use snake case**

`/api/my_api/{my_id}` ✅

`/api/my_api/{myId}` ❌

**BEWARE: path parameters should not expect values of unknown length**

Accepting very long strings in path parameters can cause issues for HTTP servers that limit the byte length of certain parts of requests.

Use a body parameter for long strings instead and consider using validation to limit a path param's length.

### Query parameters

**Should use snake case**

`/api/my_api?snake_case=true` ✅

`/api/my_api?snakeCase=false` ❌

**BEWARE: query parameters should not expect values of unknown length**

Accepting very long strings in query parameters can cause issues for HTTP servers that limit the byte length of certain parts of requests.

Use a body parameter for long strings instead and consider using validation to limit a query param's length.

### Bodies

> **Should contain camelCase keys**

Author: This is a contentious point, but imo might not be so bad to embrace. However I'm happy to revise my stance here.

if we want snake case bodies perhaps we should allow for snake case variables in JS too? What do folks think?

Reviewer 2: I think there's still some cognitive overhead when a user "knows" the field name from memory "I need to set the rule alerting frequency to 'weekly'" but then doesn't know how to type it "Is it rule
alerting_frequency or ruleAlertingFrequency??"

Other options is trying to do automatic conversion, but I think the pragmatic thing is just to allow underscore variable names in ESlint. We should keep the requirement in our styleguide that variable names should always be camelCase, but we'd prevent ESLint from babysitting developers and allow developers to make these pragmatic tradeoffs.

Alternatively as per Dario's suggestion we can only allow underscore for destructuring assignments, I'm not opposed to that, but if that still causes ESLint errors that get in the way of an engineer responsibly choosing underscore variable names for working with their API I'd lean towards just giving them more freedom and responsibility.

Author: > I'd lean towards just giving them more freedom and responsibility

++ I agree with this. One possibility, without making any drastic code changes is to leave this up to developers allowing for code like:

```typescript
    /* eslint-disable @typescript-eslint/naming-convention */
    const my_variable = 1;
    const another_variable = 2;
    /* eslint-enable @typescript-eslint/naming-convention */
    console.log(my_variable, another_variable);
```

Reviewer 3: I agree allowing snake case variables for these situations (which mostly affects devs) is probably better than  casing the request bodies differently (which affects users)

Choose sensible defaults. When uncertain, ask API callers to make informed decisions based on [documentation](#documentation).

**Should not be changed lightly**

Changing the value of a default may, in some cases, have a devastating result similar to a breaking change.

### Validation

**Runtime validation should be as narrow as feasible**

`schema.object({ id: schema.string({ minLength: 32, maxLength: 32 }) })` ✅

`schema.object({ id: schema.string() })` ❌

**It is easier to relax requirements than tighten them**

If you are in doubt, rather go with stricter validation. Make a requirement more lax if needed is never a breaking change!

### Headers

**Should not be used to specify behavior**

Outside of exceptional cases, you should always use the parameters, query parameters or body of the request to specify behavior.

### Response codes

**Should always be used as close as possible to their semantic meaning**

See the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status).

### Pagination and sorting

**Page-based request**

`GET /api/resources?page=1&size=10`

**Page-based response**

```json
{
  "items": [...],
  "total": 100,

>   "page": 1,

Author: Alternatively we can use `from` as this provides a more flexible form of pagination. E.g. Get pageNumber of resources "from" 22nd entry.

  "size": 10
}
```

**Cursor based request**

`GET /api/resources?pit_id=abc&search_after="valueA.valueB"&size=10`

Where `valueA` and `valueB` identify the last hit. See [ES docs](https://www.elastic.co/docs/reference/elasticsearch/rest-apis/paginate-search-results#search-after). Note: cursor pagination is only needed for resources exceeding 1000 instances.

**Cursor based response**

```json
{
  "items": [...],
  "total": 100,
  "pit": "abc",

>   "search_after": "valueA1.valueB1",

Author: I made this `.` join thing up. It's the two sort values we get back from ES but intended as the cursor ID. In theory we can also join the pit ID to the start to return just one `cursorID` value, but maybe there's an oversight hiding? 🤔

  "size": 10
}
```

**Pagination requires sorting**

Carefully decide a default for sorting resources. It is possible to accept a custom
set of additional values for sorting, but this may not be needed. Adding this later is a non breaking change!

Specify sorting in the query parameters like:

```
GET /api/my_resources?sort=field_name,-other_name
```

Where `field_name` is the name of the field to sort by in ascending order and `-other_name` is the name of the field to sort by in descending order.

### Filtering

`GET /api/my_resources?filter=field_name:value`

> Use [simple KQL](https://www.elastic.co/docs/explore-analyze/query-filter/languages/kql) to cover most of your filtering needs.

Author: What do folks think? Is this a good general recommendation? Should we encourage a different way of specifying filters?

<DocCallOut title="Do not break clients!">
  A public HTTP API should never cause a client to break. Not without a long (18 month) deprecation period and a ready alternative.

  Even `Technical Preview` and `Beta` HTTP APIs should consider graceful paths for changes when possible.

  Linus Torvalds famously said of Kernel development: "WE DO NOT BREAK USERSPACE!" (along with some other expletives). We should adopt this kind of rigor and empathy when working with our _public_ HTTP APIs and always prioritize finding ways to avoid breaking our APIs.
</DocCallOut>

## Internal vs Public APIs

Public HTTP APIs are those that are exposed to users and are used by plugins and external tools. They are the focus of this document.

Internal HTTP APIs are only used by code Elastic owns, like the UI. By default, when you register a new HTTP API it will be classified as `internal`. For example, this API is internal:

```typescript
router.get({ path: '/api/foos' ... }, async () => {...})
```

You can make an HTTP API public by changing the `access` setting to `public`:

```typescript
router.get({ path: '/api/foos', options: { access: 'public' } }, async () => {...})
```

> See the principle regarding [commitment](#commitment) before making APIs public!

Reviewer: Is it time to think about a Public API committee? It could be adjacent to the BC committee, with similar strict requirements for an API to make the cut as public?

## Structure and conventions

> The following are a quick reference for how to structure your HTTP APIs.

Reviewer: ```suggestion
The conventions covered below serve as a quick reference for how to structure your HTTP APIs.
```

### RESTful-like APIs

> Kibana HTTP APIs do not need to implement the full RESTful specification. We aim for a more pragmatic but still user-friendly approach.

Reviewer: If we do keep the point about `RESTful-like` design, then we need to give a bit more detail on when and how to be pragmatic and retain user-friendliness. If not and if we don't have anything concrete in mind, then maybe rephrasing it as: If in doubt, default to ESTful API design.

Author: Yeah, the point I was trying to raise is a bit nuanced. Something like: we follow a subset of RESTful practices (as opposed to full RESTful) instead of RPC or anything else. But I think this is all implied from the stuff below. I'll remove this for now!

### Path

**Use snake case**

`/api/my_api` ✅

`/api/my-api`, `/api/myApi` ❌

> **Do not contain version numbers**

Reviewer: ```suggestion
**Don't include version numbers**
```

`/api/my_api` ✅

`/api/v1/my_api` ❌

See the section on [versioning](#versioning) for more details.

**For public APIs, prefix with `/api`**

`/api/my_api` ✅

`/my_api` ❌

> **Pluralize collection names**

Reviewer: How would we differentiate working an a collection of objects in bulk?

`PUT /api/resources/{id}` ✅

`PUT /api/resource/{id}` ❌

### Methods

**GET does not accept a body**

`GET /api/resources/{id}` ✅

`GET /api/resources/{id} { body: { params } }` ❌

**POST is for creating a resource**

Sometimes `POST` can be used for an action against a resource or for actions like a `GET` but with a body. If you expect you need to support a lot of input for searching or listing resources it might be good to reserve `POST` for this purposes. For example:

```
POST /api/my_resources
{
  search_after: "<a_really_long_id>",
  size: 10000
}
```

**PUT or PATCH are for updating a resource**

> `PUT` is a bit special in that it should _create_ the resource if it does not exist. As a result it should expect a **full** resources to be provided in body of the call.

Reviewer: Optional: 
```suggestion
`PUT` is a bit special in that it should _create_ the resource if it does not exist (similar to `upsert` in elasticsearch). As a result it should expect a **full** resources to be provided in body of the call.
```

> `PATCH` is used for partial updates and should never create a resource that does not already exist.

Reviewer: ```suggestion
`PATCH` is used for partial updates and should never create a resource that does not already exist. Be very careful with `PATCH`, especially when considering Gitops/IaC use cases.
```

**DELETE is for deleting a resource**

Delete can return a simple response instead of the full resource:

```
DELETE /api/resources/{id}
# -> 204 No content
```

**Stick to simple methods**

`GET`, `POST`, `PUT`, `PATCH`, `DELETE` should cover almost all cases. If you are considering `HEAD` or some other method make sure you have a good justification for your HTTP API.

### Path parameters

**Should use snake case**

`/api/my_api/{my_id}` ✅

`/api/my_api/{myId}` ❌

**BEWARE: path parameters should not expect values of unknown length**

> Accepting very long strings in path parameters can cause issues for HTTP servers that limit the byte length of certain parts of requests.

Reviewer: Do we know what's generally safe for kibana APIs?

Use a body parameter for long strings instead and consider using validation to limit a path param's length.

### Query parameters

**Should use snake case**

`/api/my_api?snake_case=true` ✅

`/api/my_api?snakeCase=false` ❌

**BEWARE: query parameters should not expect values of unknown length**

> Accepting very long strings in query parameters can cause issues for HTTP servers that limit the byte length of certain parts of requests.

Reviewer: Duplicate(ish) of long string warning for path params. Maybe we can combine path params and query params sections, since it's pretty much identical guidence for both.

### Special data formats

**Dates**
Use [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) in UTC. This is the default representation in Node.js.

**Durations**
Can be specified in the [Elasticsearch date math](https://www.elastic.co/docs/reference/elasticsearch/rest-apis/common-options#date-math) format, for ex. `1m` for 1 minute if you want your duration to be relative to `now` where the full expression would be `now-1m`.

Alternatively, use 2 numbers in milliseconds or [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) date strings.

## Documentation

**Public HTTP APIs must have reference documentation**

See our current reference documentation [here](https://www.elastic.co/docs/api/doc/kibana/). This is compiled from our OpenAPI spec.

See <DocLink id="kibDevTutorialGeneratingOASForHTTPAPIs" text="this tutorial" /> about the code-first approach to generating OpenAPI spec available in Kibana.

## Versioning

```bash
# An example request to a versioned API
curl -v -uelastic:changeme 'http://localhost:5601/api/synthetics/monitors' \
  -H 'elastic-api-version: 2023-10-31'
```

> Kibana's public HTTP APIs are versioned with the entire Elastic organization using data-based versioning. The date indicates the last breaking change. For example: version `2023-10-31` is saying "the last breaking change was at the end of October 2023".

Author: I thought this version number would be shared between stack and serverless?

> Kibana's public HTTP APIs are versioned with the entire Elastic organization using data-based versioning. The date indicates the last breaking change. For example: version `2023-10-31` is saying "the last breaking change was at the end of October 2023".

Reviewer 3: for Kibana, yes, but not for ES
